src/org/zaproxy/zap/extension/history/PopupMenuExportURLs.java

```diff
@@ -22,131 +22,108 @@
-                } finally {
-            	    try {
-            	        fw.close();
-            	    } catch (Exception e2) {
-                    	log.warn(e2.getStackTrace(), e2);
-            	    }
-                }
-        	}
-        });
+        		performAction();
+			}
+		});
 			
 	}
-	
-    public void setExtension(ExtensionHistory extension) {
+    
+    protected void performAction() {
+		writeURLs(buildOutputSet(
+				(SiteNode) extension.getView().getSiteTreePanel().getTreeSite().getModel().getRoot()));
+	}
+
+	public void setExtension(ExtensionHistory extension) {
         this.extension = extension;
     }
-    	
-	private void exportURLs(SiteNode node, BufferedWriter writer, boolean html) {
-		
-        if (node == null) {
-            return;
-        }
-        try {
-        	if (node.getHistoryReference() != null &&
-        			(node.getHistoryReference().getHistoryType() == HistoryReference.TYPE_PROXIED ||
-        				node.getHistoryReference().getHistoryType() == HistoryReference.TYPE_ZAP_USER ||
-         				node.getHistoryReference().getHistoryType() == HistoryReference.TYPE_SPIDER)) {
-        		
-        	 	HttpMessage msg = node.getHistoryReference().getHttpMessage();
-    			if (msg != null && msg.getRequestHeader().getURI() != null) {
 
-            		writer.write(msg.getRequestHeader().getMethod());
-            		writer.write("\t");
-            		if (html) {
-                		writer.write("<a href=\"");
-            			writer.write(msg.getRequestHeader().getURI().toString());
-                		writer.write("\">");
-            			writer.write(msg.getRequestHeader().getURI().toString());
-                		writer.write("</a><br>");
-            		} else {
-            			writer.write(msg.getRequestHeader().getURI().toString());
-            		}
-            		writer.write(CRLF);
-    			}
-        	}
-			
-		} catch (HttpMalformedHeaderException e) {
-        	log.warn(e.getStackTrace(), e);
-		} catch (IOException e) {
-        	log.warn(e.getStackTrace(), e);
-		} catch (DatabaseException e) {
-        	log.warn(e.getStackTrace(), e);
+	protected SortedSet<String> buildOutputSet(SiteNode startingPoint) {
+		SortedSet<String> outputSet = new TreeSet<String>();
+		Enumeration<?> en = (startingPoint.preorderEnumeration());
+		while (en.hasMoreElements()) {
+			SiteNode node = (SiteNode) en.nextElement();
+			if (node.isRoot()) {
+				continue;
+			}
+			HistoryReference nodeHR = node.getHistoryReference();
+			if (nodeHR != null
+					&& !HistoryReference.DEFAULT_TEMPORARY_HISTORY_TYPES.contains(nodeHR.getHistoryType())) {
+            	outputSet.add(nodeHR.getURI().toString());
+            }
 		}
+		return outputSet;
+	}
 
-		@SuppressWarnings("unchecked")
-		Enumeration<SiteNode> en = node.children();
-		while (en.hasMoreElements()) {
-			exportURLs(en.nextElement(), writer, html);
+	protected void writeURLs(SortedSet<String> aSet) {
+		File file = getOutputFile();
+		if (file == null) {

```

Comment:
Maybe we should ask for the file before collecting the URLs? It would avoid collecting them if the user cancels the dialogue.